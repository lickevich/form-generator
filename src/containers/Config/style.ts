import styled from "styled-components";

export const Textarea = styled.textarea`
  width: 100%;
  height: 520px;
  padding: 5px;
  resize: none;
`;
