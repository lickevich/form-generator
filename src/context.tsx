import React, { useState } from "react";

import defaultConfig from "./data";
import { ContextType } from "./type";
import { ROUTES } from "./constants";
import { isIncludedValue } from "./utils/helpers";

export const initialState = {
  template: {
    title: "",
    items: [],
    buttons: [],
  },
  setTemplate: () => {},
  error: "",
  setError: () => {},
  config: defaultConfig,
  setConfig: () => {},
  blockedRoutesIds: [ROUTES.RESULT.ID],
  setBlockedRoutesIds: () => {},
  isResultAvailable: false,
};

export const Context = React.createContext<ContextType>(initialState);

export const ContextProvider: React.FC = ({ children }) => {
  const [template, setTemplate] = useState<ContextType["template"]>(
    initialState.template
  );
  const [error, setError] = useState<ContextType["error"]>(initialState.error);
  const [config, setConfig] = useState<ContextType["config"]>(
    initialState.config
  );
  const [blockedRoutesIds, setBlockedRoutesIds] = useState<
    ContextType["blockedRoutesIds"]
  >(initialState.blockedRoutesIds);
  const isResultAvailable: boolean = !isIncludedValue(
    blockedRoutesIds,
    ROUTES.RESULT.ID
  );

  return (
    <Context.Provider
      value={{
        template,
        setTemplate,
        error,
        setError,
        config,
        setConfig,
        blockedRoutesIds,
        setBlockedRoutesIds,
        isResultAvailable,
      }}
    >
      {children}
    </Context.Provider>
  );
};

export default ContextProvider;
