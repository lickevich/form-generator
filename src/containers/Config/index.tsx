import React, { useContext } from "react";
import { useHistory } from "react-router-dom";

import { Context } from "../../context";
import { ContextType } from "../../type";
import { Textarea } from "./style";
import { isValidConfig } from "../../utils/validaters";
import { INVALID_FORMAT, ROUTES } from "../../constants";
import Button from "../../components/Form/Button";
import Row from "../../components/Form/Row";
import ButtonsGroup from "../../components/Form/ButtonsGroup";
import { isIncludedValue } from "../../utils/helpers";

const Config: React.FC<React.ReactNode> = () => {
  const {
    config,
    setConfig,
    setTemplate,
    setError,
    setBlockedRoutesIds,
  } = useContext(Context) as ContextType;
  const history = useHistory();

  const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) =>
    setConfig(e.target.value);

  const handleSubmit = (e: React.FormEvent): void => {
    e.preventDefault();

    if (isValidConfig(config)) {
      setTemplate(JSON.parse(config));
      setError("");
      setBlockedRoutesIds((prev) =>
        prev.filter((route) => route !== ROUTES.RESULT.ID)
      );
      history.push(ROUTES.RESULT.PATH);
    } else {
      setError(INVALID_FORMAT);
      setBlockedRoutesIds((prev) =>
        isIncludedValue(prev, ROUTES.RESULT.ID)
          ? prev
          : [...prev, ROUTES.RESULT.ID]
      );
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <Row>
        <Textarea value={config} onChange={handleChange} />
      </Row>
      <ButtonsGroup>
        <Button id="" type="submit" value="Apply" />
      </ButtonsGroup>
    </form>
  );
};

export default Config;
