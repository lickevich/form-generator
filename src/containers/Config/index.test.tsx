import React from "react";
import { render, cleanup } from "@testing-library/react";

import { ContextProvider } from "../../context";
import Config from "./";

const renderWithContext = (component: JSX.Element) => {
  return {
    ...render(<ContextProvider>{component}</ContextProvider>),
  };
};

afterEach(cleanup);

describe("Config Page", () => {
  it("render the component", () => {
    const component = renderWithContext(<Config />);
    const { asFragment } = component;

    // @ts-ignore
    expect(asFragment(component)).toMatchSnapshot();
  });
});
