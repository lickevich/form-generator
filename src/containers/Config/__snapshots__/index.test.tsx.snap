// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Page render the component 1`] = `
<DocumentFragment>
  <form>
    <div
      class="sc-dkPtyc jEDgkp"
    >
      <textarea
        class="sc-bdvvaa gytgsU"
      >
        {
  "title": "Form Title",
  "items": [
    { "type": "text", "label": "textfield" },
    { "type": "number", "label": "numberfield" },
    { "type": "textarea", "label": "textarea" },
    { "type": "date", "label": "datefield" },
    {
      "type": "checkbox",
      "groupLabel": "checkbox",
      "items": [{ "label": "checkbox1" }, { "label": "checkbox2" }]
    },
    {
      "type": "radio",
      "groupLabel": "radio",
      "items": [
        { "label": "radio1", "name": "radio", "value": "1" },
        { "label": "radio2", "name": "radio", "value": "2" }
      ]
    }
  ],
  "buttons": [
    { "type": "submit", "value": "Apply" },
    { "type": "button", "value": "Ok" },
    { "type": "reset", "value": "Reset" }
  ]
}
      </textarea>
    </div>
    <div
      class="sc-hKwCoD bwaAwH"
    >
      <input
        class="sc-gsDJrp bCXitt"
        id=""
        type="submit"
        value="Apply"
      />
    </div>
  </form>
</DocumentFragment>
`;
